### DISCUSSION TITLE
How to read Env Variable inside custom SPI

### DISCUSSION THREAD

**Discussion by kabipatt** [keycloak/keycloak#22551] [Q&A]
I am struggling to read an env variable inside my SPI code. I am running KC 21.1.2 in a containerized env. Here is my code sample.

1. `KC_SPI_MY_SPI_RULESET="8443:sso.mypoc.com,80:sso.mypoc.com"`
2. My Provider id is `my-spi`.
4. I am trying to read the variable `ruleSet` by following code, but it is not reading the variable.

```
public class MySpiFactory implements HostnameProviderFactory {
:
    public void init(Config.Scope config) {
	Set<String> props = config.getPropertyNames();
	String ruleSet = config.get("ruleSet");
:
```
What I am missing here?

**Comment by justin-tay** [comment]
I'm not sure that works with environment variables.

There is a section on [Configuring providers](https://www.keycloak.org/server/configuration-provider) and it looks like you need to pass those configuration values in a particular format

```
spi-<spi-id>-<provider-id>-<property>=<value>
```

I think then you need something like

```
bin/kc.[sh|bat] start --spi-hostname-my-spi-rule-set=8443:sso.mypoc.com,80:sso.mypoc.com
```

**Comment by danielFesenmeyer** [comment]
Have you tried "ruleset"? Probably "ruleSet" expects that the env var ends with "RULE_SET" (note the underscore)?

**Comment by kabipatt** [comment]
Thank you guys!
The following variable name fixed the issue.
`KC_SPI_HOSTNAME_MY_SPI_RULESET="8443:sso.mypoc.com,80:sso.mypoc.com"`

The "HOSTNAME" is the name pf the spi provided in the base class spi `org.keycloak.urls.HostNameSpi.`  Refer KC source code. My custom provider class extend keycloakk's  `org.keycloak.urls.HostNameProvider' (This class is part of HostNameSpi eco system).
```
public class HostnameSpi implements Spi
:
  public String getName() {
    return "hostname";
  }
:
```

My custom extended class name is mySpI. 

```
public class MyHostNameProvider implements HostnameProvider {
:
  public String getName() {
    return "mySpi";
  }
: 
```
Looks like, the format must be
`KC_SPI_` + Base SPI Name + Custom Implemented/ Inherited Name + Custom Variable Name.
`KC_SPI_`  + HOSTNAME_     + MY_SPI_   +  RULESET
